mlir/lib/Dialect/Linalg/Transforms/TransposeConv2D.cpp

```diff
@@ -0,0 +1,127 @@
+
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/Passes.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/ValueRange.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "llvm/ADT/SmallVector.h"
+#include <memory>
+#include <numeric>
+
+namespace mlir {
+#define GEN_PASS_DEF_LINALGTRANSPOSECONV2D
+#include "mlir/Dialect/Linalg/Passes.h.inc"
+} // namespace mlir
+
+using namespace mlir;
+
+namespace {
+// Convolution converter that applies the following rewrite:
+//
+// clang-format off
+// Before:
+//
+//   %0 = linalg.conv_2d_nhwc_fhwc {dilations = dense<1> : tensor<2xi64>,
+//                                               strides = dense<2> : tensor<2xi64>}
+//      ins (%input, %filter: tensor<1x4x4x6xf32>, tensor<8x2x2x6xf32>)
+//     outs (%init: tensor<1x2x2x8xf32>) -> tensor<1x2x2x8xf32>
+//
+// After:
+//
+//    %cst = arith.constant 0.000000e+00 : f32
+//    %0 = tensor.empty() : tensor<2x2x6x8xf32>
+//    %1 = linalg.fill ins(%cst : f32) outs(%0 : tensor<2x2x6x8xf32>) -> tensor<2x2x6x8xf32>
+//    %transposed = linalg.transpose ins(%arg1 : tensor<8x2x2x6xf32>) outs(%1 : tensor<2x2x6x8xf32>)
+//                  permutation = [1, 2, 3, 0]
+//    %2 = linalg.conv_2d_nhwc_hwcf {dilations = dense<1> : tensor<2xi64>, strides = dense<2> : tensor<2xi64>}
+//         ins(%arg0, %transposed : tensor<1x4x4x6xf32>, tensor<2x2x6x8xf32>) outs(%arg2 : tensor<1x2x2x8xf32>)
+//         -> tensor<1x2x2x8xf32>
+//
+// clang-format on
+// with an analogous example for the quantized case.
+template <typename FHWCConvOp, typename HWCFConvOp>
+class ConvConverter : public OpRewritePattern<FHWCConvOp> {
+public:
+  using OpRewritePattern<FHWCConvOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(FHWCConvOp op,
+                                PatternRewriter &rewriter) const final {
+    // Construct a permutation of the filter tensor dimensions. For a 2D
+    // convolution this will be known statically as [1, 2, 3, 0].
+    auto resultTy = cast<ShapedType>(op->getResult(0).getType());
+    auto filterPerm = SmallVector<int64_t>({1, 2, 3, 0});
+
+    // Create the type for the transposed filter tensor since this will be
+    // different from the original filter type.
+    auto filter = op->getOperand(1);
+    auto filterTy = cast<ShapedType>(filter.getType());
+    auto newfilterShape = SmallVector<int64_t>(filterPerm.size());
+    std::generate(std::begin(newfilterShape), std::end(newfilterShape),
+                  [dim = 0, &filterTy, &filterPerm]() mutable {
+                    return filterTy.getShape()[filterPerm[dim++]];
+                  });
+    auto newFilterTy =
+        RankedTensorType::get(newfilterShape, filterTy.getElementType());
```

Comment:
I think these ops can also take memrefs, would this transform be correct in that case?

Reply:
It's sorta unclear from the docs: https://mlir.llvm.org/docs/Dialects/Linalg/#linalgconv_2d_nhwc_fhwc-linalgconv2dnhwcfhwcop, but it says for the output tensor at least _variadic of ranked tensor of any type values_ which makes it sound like it can only accept tensors?

I tried to lit test this locally and it failed to verify the IR saying: `op result #0 must be ranked tensor of any type values, but got memref....`

Reply:
Oh I see you construct the `RankedTensorType`:
```c++
auto newFilterTy =
        RankedTensorType::get(newfilterShape, filterTy.getElementType());
```
but that type is never really used:
```c++
    auto emptyTensor = rewriter.create<tensor::EmptyOp>(
        loc, newFilterTy.getShape(), resultETy);
```
^ only the shape of the type is passed here. So you could just change that to:
```c++
    auto emptyTensor = rewriter.create<tensor::EmptyOp>(
        loc, newfilterShape, resultETy);
```

Reply:
Also nit: `newfilterShape` -> `newFilterShape`

Reply:
Ah yep, good point I'll just remove the `newFilterTy` altogether then :D

Reply:
Also, it looks like:
```
func.func @conv_2d_nhwc_fhwc(%arg0:
  memref<?x?x?x?xf32>,
  %arg1: memref<?x?x?x?xf32>,
  %arg2: memref<?x?x?x?xf32>
) -> memref<?x?x?x?xf32> {
  linalg.conv_2d_nhwc_fhwc {dilations = dense<1> : tensor<2xi64>, strides = dense<1> : tensor<2xi64>}
    ins(%arg0, %arg1 : memref<?x?x?x?xf32>, memref<?x?x?x?xf32>)
    outs(%arg2 : memref<?x?x?x?xf32>)
  return %arg2 : memref<?x?x?x?xf32>
}
```
round trips fine, but:
```
func.func @conv_2d_nhwc_fhwc(%arg0:
  memref<?x?x?x?xf32>,
  %arg1: memref<?x?x?x?xf32>,
  %arg2: memref<?x?x?x?xf32>
) -> memref<?x?x?x?xf32> {
  linalg.conv_2d_nhwc_fhwc {dilations = dense<1> : tensor<2xi64>, strides = dense<1> : tensor<2xi64>}
    ins(%arg0, %arg1 : memref<?x?x?x?xf32>, memref<?x?x?x?xf32>)
    outs(%arg2 : memref<?x?x?x?xf32>) -> <?x?x?x?xf32>
  return %arg2 : memref<?x?x?x?xf32>
}
```
fails to validate (note the trailing return type on the conv).

Looking at the other LinAlg tests it seems like the trailing return type is valid syntax for the operator, so I'm inclined to say this a bug.

For now I'll omit it from my lit test so we can get this merged but maybe I should file a bug for this?

Reply:
So I've tried adding a test for this and it seems the `linalg.fill` operation doesn't support memrefs. I guess the correct thing to do here would be to handle memrefs with a different operation, but that might be something we could look at adding support for in the future ratht than doing as part of this PR?

Reply:
`linalg.fill` does support memrefs, I've used it on them in my own tests:
```
linalg.fill ins(%value : f32) outs(%my_memref :memref<1x2x?xf32>)
```
The result is unused as the memref since the fill directly updates the memory the memref points to.

Reply:
Oops sorry for the delay, this required more changes than I thought. I've updated the pass to handle both tensors and memrefs now. This requires a bit of branching to accomdate the different dialects e.g. tensor.empty vs. memref.alloc and the passing by out parameter that seems to be required for the memrefs. I've also simplified the transformation slightly - I realized you don't actually need a tensor/memref of zeros since you just overwrite it anyway so you can allocate it then pass it directly to the transpose.